### ISSUE TITLE
refactor: migrate VAE/conditioner/SDXL composites to multi-host CompiledModelHost so they can consume Tensors ChainAsync (gated on AiDotNet.Tensors#296)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1272]
## Context

Downstream consolidation issue gated on [`ooples/AiDotNet.Tensors#296`](#) (to be cross-linked once filed). The Tensors-side issue redesigns `ICompiledPlan<T>` to be genuinely async (`ExecuteAsync` + `ChainAsync` over per-engine streams / threadpools), drops the misnamed sync `ThenAsync`, and adds multi-input chain support. Once that lands and a Tensors NuGet release ships, AiDotNet has structural work to do before any consumer can use it.

This issue tracks the AiDotNet-side refactor that closes the consumption gap. It supersedes [`ooples/AiDotNet#1157`](https://github.com/ooples/AiDotNet/issues/1157) (Use ICompiledPlan.ThenAsync for multi-stage model chaining), whose original premise was refuted by audit — see #1157's close comment for the full audit findings.

## What's broken today (AiDotNet side)

The audit confirmed AiDotNet has **zero** usable multi-stage compiled-host pipelines today:

- Only TWO classes own a `CompiledModelHost<T>` in all of `src/`: `NeuralNetworkBase<T>` and `NoisePredictorBase<T>`. Each is a single-host monolith.
- `VAEModelBase<T>` does not own a `CompiledModelHost`. `VAEEncoder<T>` and `VAEDecoder<T>` are `LayerBase<T>` instances called via `Forward(input)` directly — no compilation infrastructure on the VAE path
- `SDXLModel<T>` (the most complex composite in the codebase) holds 4 sub-models (`_conditioner1`, `_conditioner2`, `_unet`, `_vae`):
  - `_unet` is a `UNetNoisePredictor<T>` → has its own `CompiledModelHost` ✅
  - `_vae` is a `StandardVAE<T>` → does **not** have a `CompiledModelHost` ❌
  - `_conditioner1`, `_conditioner2` are `IConditioningModule<T>` → ownership audit needed but currently no `CompiledModelHost` reachable on the conditioner code path

So even after the Tensors-side `ChainAsync` ships, AiDotNet has nothing to chain. Chaining requires ≥2 separately-compiled hosts in the same Predict path. Today: 0.

### 1. Promote `VAEEncoder<T>` / `VAEDecoder<T>` from `LayerBase<T>` to `NeuralNetworkBase<T>`

Current shape:

```csharp
public class VAEEncoder<T> : LayerBase<T>      // → no CompiledModelHost
public class VAEDecoder<T> : LayerBase<T>      // → no CompiledModelHost
```

Target shape (mirrors the existing pattern in `NoisePredictorBase`):

```csharp
public class VAEEncoder<T> : NeuralNetworkBase<T>   // owns its own _compileHost
public class VAEDecoder<T> : NeuralNetworkBase<T>   // owns its own _compileHost
```

Each subclass implements `InitializeLayers` to register the existing layer composition (`_inputConv`, `_downBlocks`, `_midBlocks`, etc.) under the `NeuralNetworkBase` umbrella. Existing call sites in `VAEModelBase.Predict / Encode / Decode` continue to work — they call `Forward` on a `LayerBase` today; switching to `Predict` on a `NeuralNetworkBase` is a trivial rename + `using NoGradScope` boundary that already exists in `Predict` overrides.

Reference impl: `NoisePredictorBase<T>` already shows the pattern of "wrap a sub-network in `CompiledModelHost`, expose `Predict`/`PredictNoise`". Same shape, same trace-and-compile flow.

### 2. Audit `IConditioningModule<T>` implementations for compile-host ownership

Concrete conditioners (CLIP-style text encoders, T5 text encoders) are called once per generation in `SDXLModel.Generate` etc. Today they go through whatever forward path the implementation provides. Bring them under `NeuralNetworkBase<T>` so the conditioner pre-step compiles to a plan that can be the FIRST stage of a chain.

Subtask: file an inventory of `IConditioningModule<T>` implementations in `src/Diffusion/Conditioning/`, audit each for `CompiledModelHost` reachability, refactor those that aren't reachable.

### 3. Add `ChainedCompiledModelHost<T>` public helper

A wrapper that takes 2+ `CompiledModelHost<T>` instances + a chain definition and routes through Tensors' `ChainAsync`. Lives in `src/NeuralNetworks/`:

```csharp
public sealed class ChainedCompiledModelHost<T> : IDisposable
{
    public ChainedCompiledModelHost(params CompiledModelHost<T>[] stages);

    /// Synchronous predict — falls back to per-stage Predict when the
    /// caller doesn't want async. Useful for callers that already block
    /// in a tight loop.
    public Tensor<T> Predict(Tensor<T> input, int structureVersion, ...);

    /// Async predict — uses Tensors' ChainAsync so stages overlap on the
    /// engine's stream / threadpool. THIS is the path the diffusion
    /// pipeline should take.
    public ValueTask<Tensor<T>> PredictAsync(
        Tensor<T> input, int structureVersion,
        CancellationToken ct = default);

    /// Multi-input chain: stage k consumes (output_of_(k-1), additional_inputs[k]).
    /// Maps to Tensors' ChainAsync(next, nextInputSlot, ct).
    public ValueTask<Tensor<T>> PredictAsync(
        Tensor<T> primaryInput,
        Tensor<T>[] sideInputsPerStage,    // [stage_index][side_input_index]
        int structureVersion,
        CancellationToken ct = default);

    public void Dispose();
}
```

This is the AiDotNet-public API. Composite models call `new ChainedCompiledModelHost<T>(conditioner1Host, unetHost, vaeDecoderHost)` and dispatch `PredictAsync` to inherit all the Tensors-side stream/overlap behavior.

### 4. Rewire `SDXLModel.Generate` (the canonical 4-stage composite)

Today's structure (paraphrased):

```csharp
public Tensor<T> Generate(...)
{
    var emb1 = _conditioner1.EncodeText(tokens1);
    var emb2 = _conditioner2.EncodeText(tokens2);   // sequential, no chain
    Tensor<T> latent = SampleNoise(...);
    foreach (var t in _scheduler.Timesteps)         // 50 sync sub-net Predict calls
        latent = _unet.PredictNoise(latent, t, emb1, emb2);
    return _vae.Decode(latent);
}
```

Target structure:

```csharp
public async ValueTask<Tensor<T>> GenerateAsync(...)
{
    // Concurrent conditioner pre-step — emb1 and emb2 are independent,
    // they should run on different streams.
    var conditioning = await Task.WhenAll(
        _conditioner1Host.PredictAsync(tokens1, ...).AsTask(),
        _conditioner2Host.PredictAsync(tokens2, ...).AsTask())
        .ConfigureAwait(false);

    Tensor<T> latent = SampleNoise(...);
    foreach (var t in _scheduler.Timesteps)
    {
        // Async per-step so the next step's host-side work (timestep
        // embedding, scheduler.Step, sample tensor refresh) can start
        // before the GPU finishes the current step's tail kernels.
        latent = await _unetHost.PredictAsync(
            latent, ..., sideInputs: conditioning, ...).ConfigureAwait(false);
    }

    // Final VAE decode — fully async so any host work after Generate
    // returns (e.g. image-encode) can overlap with the GPU tail.
    return await _vaeDecodeHost.PredictAsync(latent, ...).ConfigureAwait(false);
}
```

Sync `Generate` stays as a thin wrapper that calls `GenerateAsync(...).GetAwaiter().GetResult()` for callers that don't want to migrate to async. New code should use the async path.

### 5. Wire conditioner output → noise-predictor as multi-input ChainAsync (when stable)

The most valuable chain in latent diffusion is `text_conditioner_output → cross_attention_noise_predictor`. This is the chain shape that Tensors' multi-input `ChainAsync` (issue's acceptance criterion #8) is designed for. Once the Tensors API ships, replace the per-step manual side-input plumbing in step 4 with a stitched `ChainAsync` so the conditioner's output flows directly into the noise predictor's cross-attention input without ever materializing on the host.

This is the highest-leverage AiDotNet consumer of the new API. SDXL inference latency benchmarks (head-to-head vs PyTorch SDXL) gate the success of the whole effort.

### 6. Diffusion denoising-loop pipelining

Even with per-step `PredictAsync`, the 50-step loop is still serial — step N+1 can't start before step N's output is known (it IS the input). But the host-side scheduler work (`scheduler.Step`, `noisePredVec.CopyFrom`, RNG advance for variance schedules) CAN run concurrently with the GPU's tail kernels of step N. Pipelining this saves per-step host overhead which adds up over 50 steps.

## Apples-to-apples PyTorch acceptance criteria

The Tensors-side issue establishes the `ChainAsync` perf bar (≤0.95× of PyTorch `nn.Sequential` on N-batch throughput, ≤1.10× on single-batch latency, etc.). The AiDotNet downstream PR's acceptance criteria layer on top of those:

1. **SDXL end-to-end latency**: a hand-rolled SDXL benchmark in `AiDotNetBenchmarkTests/Diffusion/SDXLEndToEndBenchmark.cs` measures full image generation (text → 50-step denoise → VAE decode → fp32 image tensor) against an equivalent PyTorch SDXL pipeline (TorchSharp wrapper or pip-installed `diffusers` via subprocess + JSON results parsing). Tensors must measure **≤ 1.10×** PyTorch's wall time at fp32, CPU, batch_size=1, 50 inference steps, 256×256 output. (The 10% window accounts for libtorch's extensive CPU autotuning.)
2. **VAE encode + decode round-trip**: separate bench measures VAE-only round-trip on a 512×512 RGB input. Tensors `VAEModelBase.Predict` (going through the new `ChainedCompiledModelHost`) must be **≤ 1.05×** PyTorch VAE.
3. **Diffusion 50-step throughput at batch_size=4**: throughput-mode bench with PyTorch torch.compile-warmup vs Tensors warm cache. Tensors must be **≤ 0.95×** PyTorch (a head-to-head win on the throughput cell).
4. **Conditioner pre-step concurrency**: the dual-conditioner SDXL path (CLIP-G + CLIP-L running concurrently) measures < 1.5× single-conditioner latency (i.e., the second conditioner runs mostly overlapped). PyTorch's `torch.cuda.stream` model achieves ~1.1× on this; we should be within striking distance.
5. **Memory parity**: Tensors `GenerateAsync` allocates **≤ 1 byte / step** in the denoising loop (i.e., the per-step host-side overhead is constant, not proportional to step count). PyTorch achieves this via `torch.no_grad()` + pre-allocated buffers; we already have `NoGradScope` and pre-allocated `noisePredVec` so this is mostly mechanical.
6. **No regression** on existing single-stage `NeuralNetworkBase.Predict` benchmarks: existing single-host paths measure within 1% of pre-PR baseline.
7. **ABI stability**: `VAEEncoder.Forward(Tensor<T>)` continues to work for callers that aren't on the chained path — the `LayerBase → NeuralNetworkBase` promotion is additive (existing callers use `Forward`, new callers use `Predict` / `PredictAsync`).

The PR body must include filled-in benchmark results for each criterion with measured Mean / StdDev / Ratio-vs-PyTorch and an explicit pass/fail call per criterion.

## Sequencing — gating on Tensors

This issue cannot start implementation until the Tensors-side issue ships a NuGet release with:
- `ICompiledPlan<T>.ExecuteAsync(CancellationToken)` returning `ValueTask<Tensor<T>>`
- `ICompiledPlan<T>.ChainAsync(ICompiledPlan<T>, CancellationToken)` returning `ValueTask<Tensor<T>>`
- Multi-input `ChainAsync(ICompiledPlan<T>, int slot, CancellationToken)`
- `ThenAsync` deprecated with `[Obsolete]`
- Per-engine `IExecutionStream<T>` implementation on `CudaBackend` and `CpuEngine`

Until then this issue stays in `Status: Blocked` with a link to the Tensors PR.

## Closes / supersedes

- [`ooples/AiDotNet#1157`](https://github.com/ooples/AiDotNet/issues/1157)
